driver/multilabelsoftmarginloss_driver.hpp

```diff
@@ -0,0 +1,558 @@
+            std::vector<int> o_lens = {N};
+            SetTensorNd(oDesc, o_lens, data_type);
+        }
+        else
+        {
+            std::vector<int> o_lens = {1};
+            SetTensorNd(oDesc, o_lens, data_type);
+        }
+    }
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+std::vector<int> MultilabelSoftMarginLossDriver<Tgpu, Tref>::ParseInputList(std::string input_str)
+{
+    std::vector<int> lengths;
+    std::size_t pos = 0;
+    std::size_t new_pos;
+
+    new_pos = input_str.find(',', pos);
+    while(new_pos != std::string::npos)
+    {
+        std::string slice_str = input_str.substr(pos, new_pos - pos);
+
+        int len = std::stoi(slice_str);
+
+        lengths.push_back(len);
+
+        pos     = new_pos + 1;
+        new_pos = input_str.find(',', pos);
+    };
+
+    std::string slice_str = input_str.substr(pos);
+    int len               = std::stoi(slice_str);
+
+    lengths.push_back(len);
+
+    return (lengths);
+}
+
+template <typename Tgpu, typename Tref>
+int MultilabelSoftMarginLossDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
+{
+    uint32_t ctx = 0;
+
+    // for unpacked tensor, we need to use GetTensorSpace instead of GetTensorSize
+    size_t i_sz = GetTensorSpace(iDesc);
+    size_t t_sz = GetTensorSpace(tDesc);
+    size_t w_sz = GetTensorSpace(wDesc);
+    i_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, i_sz, sizeof(Tgpu)));
+    t_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, t_sz, sizeof(Tgpu)));
+    w_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, w_sz, sizeof(Tgpu)));
+    I           = std::vector<Tgpu>(i_sz);
+    T           = std::vector<Tgpu>(t_sz);
+    W           = std::vector<Tgpu>(w_sz);
+    for(int i = 0; i < i_sz; i++)
+    {
+        I[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(-1), static_cast<Tgpu>(1));
+    }
+    // 0 or 1
+    for(int i = 0; i < t_sz; i++)
+    {
+        T[i] = prng::gen_A_to_B<int32_t>(static_cast<int32_t>(0), static_cast<int32_t>(2));
+    }
+    for(int i = 0; i < w_sz; i++)
+    {
+        W[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(-1), static_cast<Tgpu>(1));
+    }
+
+    if(i_dev->ToGPU(GetStream(), I.data()) != 0)
+        std::cerr << "Error copying (I) to GPU, size: " << i_dev->GetSize() << std::endl;
+
+    if(t_dev->ToGPU(GetStream(), T.data()) != 0)
+        std::cerr << "Error copying (T) to GPU, size: " << t_dev->GetSize() << std::endl;
+
+    if(w_dev->ToGPU(GetStream(), W.data()) != 0)
+        std::cerr << "Error copying (W) to GPU, size: " << w_dev->GetSize() << std::endl;
+
+    if(forw == 0 || forw == 1)
+    {
+        size_t o_sz = GetTensorSpace(oDesc);
+        if(divisor != 0)
+        {
+            miopenGetMultilabelSoftMarginLossForwardWorkspaceSize(
+                GetHandle(),
+                iDesc,
+                tDesc,
+                wDesc,
+                oDesc,
+                divisor == 1 ? MIOPEN_LOSS_REDUCTION_SUM : MIOPEN_LOSS_REDUCTION_MEAN,
```

Comment:
I think you should assign a private var "miopenLossReductionMode_t reductionMode", initialize the reductionMode based on the command line arg then pass the reductionMode to the function. Same for other functions call with the reduction mode

Reply:
fixed in [this commit](https://github.com/moreh-dev/MIOpen/pull/33/commits/7d2ecff0b467e988d82ee7331ef38b58b9091a7e)